src/frontend/Typechecker.ml

```diff
@@ -612,22 +606,31 @@ and check_reduce_sum ~is_cond_dist loc cf tenv id tes =
           mandatory_fun_args UReal arg_types
     | _ -> basic_mismatch () in
   match tes with
-  | {expr= Variable fname; _} :: remaining_es -> (
-    match find_matching_first_order_fn tenv (matching remaining_es) fname with
-    | SignatureMismatch.UniqueMatch (ftype, promotions) ->
-        (* a valid signature exists *)
-        let tes = make_function_variable cf loc fname ftype :: remaining_es in
-        mk_fun_app ~is_cond_dist ~loc (StanLib FnPlain) id
-          (Promotion.promote_list tes promotions)
-          ~type_:UnsizedType.UReal
-    | AmbiguousMatch ps ->
-        Semantic_error.ambiguous_function_promotion loc fname.name None ps
-        |> error
-    | SignatureErrors (expected_args, err) ->
-        Semantic_error.illtyped_reduce_sum loc id.name
-          (List.map ~f:type_of_expr_typed tes)
-          expected_args err
-        |> error )
+  | {expr= Variable fname; _}
+    :: ({emeta= {type_= slice_type; _}; _} :: _ as remaining_es) -> (
+      let slice_type, n = UnsizedType.unwind_array_type slice_type in
+      if n = 0 then (
```

Comment:
Does this implicitly allow larger dimensionalities than before (I assume the C++ doesn't care, just checking)?

Reply:
Yes. A real array with 9 dimensions compiled when I tested it. I don't think there's any explicit limits in the C++, it's just templates that can recurse to any depth. Stanc3 limit was probably for `Stan_math_signatures.pretty_print_all_math_sigs`.